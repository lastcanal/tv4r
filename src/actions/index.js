import { batch } from 'react-redux'

import { extractPosts, filterPosts, matchRedditPath } from '../helpers'
import reddit from '../helpers/reddit'

import {
  SELECT_SUBREDDIT,
  INVALIDATE_SUBREDDIT,
  REMOVE_SUBREDDIT,
  REQUEST_POSTS,
  RECEIVE_POSTS,
  RECEIVE_POSTS_ERROR,
  SELECT_POST,
  NEXT_POST,
  PREVIOUS_POST,
  MEDIA_FALLBACK,
  REQUEST_COMMENTS,
  RECEIVE_COMMENTS,
  RECEIVE_COMMENTS_ERROR,
  TOGGLE_FULLSCREEN,
  TOGGLE_PLAY,
  TOGGLE_AUTOPLAY,
  TOGGLE_THEME_MODE,
  PLAYER_VOLUME_UP,
  PLAYER_VOLUME_DOWN,
  PLAYER_SCAN_FORWARDS,
  PLAYER_SCAN_BACKWARDS,
  PLAYER_SCAN_ACK,
  PLAYER_JUMP_TO,
  PLAYER_JUMP_ACK,
} from '../constants'

export const selectSubreddit = subreddit => ({
  type: SELECT_SUBREDDIT,
  subreddit,
})

export const invalidateSubreddit = subreddit => ({
  type: INVALIDATE_SUBREDDIT,
  subreddit,
})

export const removeSubreddit = subreddit => ({
  type: REMOVE_SUBREDDIT,
  subreddit,
})

export const requestPosts = subreddit => ({
  type: REQUEST_POSTS,
  subreddit,
})

export const receivePosts = (subreddit, json) => ({
  type: RECEIVE_POSTS,
  subreddit,
  posts: filterPosts(extractPosts(json)),
  error: null,
  receivedAt: Date.now(),
})

export const receivePostsError = (subreddit, error) => ({
  type: RECEIVE_POSTS_ERROR,
  subreddit,
  error,
})

export const selectPost = (post, index) => ({
  type: SELECT_POST,
  post,
  index,
})

export const nextPost = posts => ({
  type: NEXT_POST,
  posts,
})

export const previousPost = posts => ({
  type: PREVIOUS_POST,
  posts,
})

export const mediaFallback = () => ({
  type: MEDIA_FALLBACK,
})

const fetchPosts = subreddit => dispatch => {
  dispatch(requestPosts(subreddit))
  return reddit
    .fetchPosts(subreddit)
    .then(response => response.json())
    .then(response => dispatch(receivePosts(subreddit, response)))
    .catch(error => dispatch(receivePostsError(subreddit, error)))
}

const shouldFetchPosts = ({ postsBySubreddit }, subreddit) => {
  const posts = postsBySubreddit[subreddit]
  if (!posts) {
    return true
  }
  if (posts.isFetching) {
    return false
  }
  return posts.didInvalidate
}

export const fetchPostsIfNeeded = subreddit => (dispatch, getState) => {
  if (shouldFetchPosts(getState(), subreddit)) {
    return dispatch(fetchPosts(subreddit))
  }
}

const shouldInvalidateSubreddit = ({ router, postsBySubreddit }, subreddit) => {
  const match = matchRedditPath(router.location.pathname)
  const posts = postsBySubreddit[subreddit]
  return (
    !posts ||
    posts.items.length === 0 ||
    (match && match.subreddit === subreddit)
  )
}

export const invalidateSubredditIfNeeded = subreddit => (
  dispatch,
  getState,
) => {
  const newSubreddit = subreddit.toLowerCase()
  if (shouldInvalidateSubreddit(getState(), newSubreddit)) {
    return dispatch(invalidateSubreddit(newSubreddit))
  }
}

export const loadSubreddit = subreddit => (dispatch) => {
  batch(() => {
    dispatch(selectSubreddit(subreddit))
    dispatch(invalidateSubredditIfNeeded(subreddit))
    dispatch(fetchPostsIfNeeded(subreddit))
  })
}

const requestComments = (post, subreddit) => ({
  type: REQUEST_COMMENTS,
  post,
  subreddit,
})

const receiveComments = (post, subreddit, response) => ({
  type: RECEIVE_COMMENTS,
  post,
  subreddit,
  comments: response,
})

const receiveCommentsError = (post, subreddit, error) => ({
  type: RECEIVE_COMMENTS_ERROR,
  post,
  subreddit,
  error,
})

const fetchComments = (post, selectedSubreddit) => dispatch => {
  dispatch(requestComments(post, selectedSubreddit))
  return reddit
    .fetchPost(post.permalink)
    .then(response => response.json())
    .catch(error =>
      dispatch(receiveCommentsError(post, selectedSubreddit, error)),
    )
    .then(response =>
      dispatch(receiveComments(post, selectedSubreddit, response)),
    )
}

export const fetchCommentsIfNeeded = () => (dispatch, getState) => {
  const { postsBySubreddit, selectedSubreddit } = getState()
  const { index } = postsBySubreddit.cursor
  const posts = postsBySubreddit[selectedSubreddit].items
  if (posts[index] && !posts[index].comments) {
    dispatch(fetchComments(posts[index], selectedSubreddit))
  }
}

export const configToggleFullscreen = () => ({
  type: TOGGLE_FULLSCREEN,
})

export const configToggleAutoplay = () => ({
  type: TOGGLE_AUTOPLAY,
})

export const configTogglePlay = () => ({
  type: TOGGLE_PLAY,
})

export const configToggleThemeMode = () => ({
  type: TOGGLE_THEME_MODE,
})

export const playerScanForwards = seconds => ({
  type: PLAYER_SCAN_FORWARDS,
  seconds,
})

export const playerScanBackwards = seconds => ({
  type: PLAYER_SCAN_BACKWARDS,
  seconds,
})

export const playerScanAck = () => ({
  type: PLAYER_SCAN_ACK,
})

export const playerJumpTo = (number) => ({
  type: PLAYER_JUMP_TO,
  percentage: (number / 10),
})

export const playerJumpAck = () => ({
  type: PLAYER_JUMP_ACK,
})

export const playerVolumeDown = () => ({
  type: PLAYER_VOLUME_DOWN,
})

export const playerVolumeUp = () => ({
  type: PLAYER_VOLUME_UP,
})

export const handleKeyboardAction = event => (dispatch, getState) => {
  const { postsBySubreddit, selectedSubreddit } = getState()
  const posts = postsBySubreddit[selectedSubreddit] &&
    postsBySubreddit[selectedSubreddit].items
  switch (event.key) {
    case '.':
    case '>':
      return dispatch(nextPost(posts))
    case 'N':
    case 'n':
      return event.shiftKey ? dispatch(nextPost(posts)) : void (0)
    case ',':
    case '<':
      return dispatch(previousPost(posts))
    case 'P':
    case 'p':
      return event.shiftKey ? dispatch(previousPost(posts)) : void (0)
    case ' ':
    case 'Enter':
    case 'k':
      return dispatch(configTogglePlay())
    case 'a':
      return dispatch(configToggleAutoplay())
    case 'f':
      return dispatch(configToggleFullscreen())
    case 'ArrowLeft':
      return dispatch(playerScanBackwards(5))
    case 'ArrowRight':
      return dispatch(playerScanForwards(5))
    case 'ArrowUp':
      return dispatch(playerVolumeUp())
    case 'ArrowDown':
      return dispatch(playerVolumeDown())
    default:
      const jumpTo = parseInt(event.key)
      if (jumpTo) {
        return dispatch(playerJumpTo(jumpTo))
      } else {
        return
      }
  }
}
