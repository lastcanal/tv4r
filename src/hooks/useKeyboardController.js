import { useEffect } from 'react'

import {
  nextPost,
  previousPost,
  configTogglePlay,
  configToggleAutoAdvance,
  configEnableFullscreen,
  configDisableFullscreen,
  playerScanBackwards,
  playerScanForwards,
  playerVolumeUp,
  playerVolumeDown,
  playerJumpTo,
} from '../actions'


export const handleKeyboardAction = event => (dispatch, getState) => {
  const { config } = getState()
  if (!config.keyboardControls) return
  if (event.metaKey) return
  switch (event.key) {
    case '.':
    case '>':
      return dispatch(nextPost())
    case 'N':
    case 'n':
      return event.shiftKey ? dispatch(nextPost()) : void (0)
    case ',':
    case '<':
      return dispatch(previousPost())
    case 'P':
    case 'p':
      return event.shiftKey ? dispatch(previousPost()) : void (0)
    case ' ':
    case 'Space':
    case 'Enter':
    case 'k':
      if (!event.shiftKey) event.preventDefault()
      return dispatch(configTogglePlay())
    case 'a':
      return dispatch(configToggleAutoAdvance())
    case 'f':
      return dispatch(configEnableFullscreen())
    case 'Escape':
      return dispatch(configDisableFullscreen())
    case 'ArrowLeft':
      return dispatch(playerScanBackwards(5))
    case 'ArrowRight':
      return dispatch(playerScanForwards(10))
    case 'ArrowUp':
      if (!event.shiftKey) event.preventDefault()
      return dispatch(playerVolumeUp())
    case 'ArrowDown':
      if (!event.shiftKey) event.preventDefault()
      return dispatch(playerVolumeDown())
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
      const jumpTo = parseInt(event.key, 10)
      return dispatch(playerJumpTo(jumpTo))
    case 'End':
      // if (!event.shiftKey) event.preventDefault()
      return dispatch(playerJumpTo(10))
    default:
      return
  }
}

const useKeyboardController = ({ dispatch }) => {
  const handleKeyDown = (event) => {
    dispatch(handleKeyboardAction(event))
  }

  useEffect(() => {
    document.body.addEventListener('keydown', handleKeyDown)
    return () => {
      document.body.removeEventListener('keydown', handleKeyDown)
    }
  }, [])
}

export default useKeyboardController
