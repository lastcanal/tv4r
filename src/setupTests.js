import React from 'react'
import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'
import fetch, { GlobalWithFetchMock } from 'jest-fetch-mock'
import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16' // React 16 Enzyme adapter
import { routerMiddleware } from 'connected-react-router'
import fc from 'fast-check'

// https://github.com/jsdom/jsdom/pull/2626
Object.defineProperty(window, 'scrollTo', { value: () => {}, writable: true })

Enzyme.configure({ adapter: new Adapter() }) // Make Enzyme functions available in all test files without importing

export const history = {
  location: {},
  action: 'POP',
  push: jest.fn(),
  replace: jest.fn(),
  go: jest.fn(),
  goBack: jest.fn(),
  goForward: jest.fn(),
  listen: jest.fn(),
}
const middleware = [thunk, routerMiddleware(history)]
const mockStore = configureMockStore(middleware)

const makeStore = (extra = {}) => {
  return mockStore({
    postsBySubreddit: {
      foo: {
        scope: 'new',
      },
    },
    selectedPost: {},
    selectedSubreddit: 'foo',
    subreddits: ['foo'],
    config: {
      themeMode: 'dark',
      isFullscreen: false,
      isPlaying: false,
      isAutoplay: false,
    },
    router: {
      history,
      location: {
        pathname: '/',
        search: '',
        hash: '',
      },
      action: 'POP',
    },
    ...extra,
  })
}

const customGlobal: GlobalWithFetchMock = global
customGlobal.fetch = fetch
customGlobal.fetchMock = customGlobal.fetch
customGlobal.shallow = shallow
customGlobal.render = render
customGlobal.mount = mount
customGlobal.React = React
customGlobal.makeStore = makeStore
// TODO: rename mockStore -> makeStore
customGlobal.mockStore = makeStore
customGlobal.fc = fc
