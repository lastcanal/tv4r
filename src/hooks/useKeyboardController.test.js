import { Provider } from 'react-redux'
import { ConnectedRouter } from 'connected-react-router'

import useKeyboardController, { handleKeyboardAction } from './useKeyboardController'
import { history } from '../setupTests'

import * as actions from '../actions'

describe('useKeyboardController', () => {

  it('handle heydown events', () => {
    const Wrapped = ({ dispatch }) => {
      useKeyboardController(store)
      return <div />
    }

    const store = makeStore()
    const wrapper = mount(
      <Provider store={store}>
        <ConnectedRouter history={history} hashType="noslash">
          <Wrapped dispatch={store.dispatch}/>
        </ConnectedRouter>
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })


  it('should only work if enabled', () => {
    const store = makeStore({ config: { keyboardControls: false }})
    store.clearActions()
    store.dispatch(handleKeyboardAction({ key: '.' }))
    expect(store.getActions()[0]).toEqual(undefined)
  })

  it('should not work if metakey pressed', () => {
    const store = makeStore({ config: { keyboardControls: false }})
    store.clearActions()
    store.dispatch(handleKeyboardAction({ key: '.', metaKey: true }))
    expect(store.getActions()[0]).toEqual(undefined)
  })

  const testKeyboardAction = (key, action, extra = {}) => {
    const store = makeStore({ config: { keyboardControls: true }})
    store.clearActions()
    store.dispatch(handleKeyboardAction({ ...extra, key }))
    store.dispatch(action)
    const [a, b] = store.getActions()
    expect(a).toEqual(b)
  }

  it('should goto nextPost ', () => {
    testKeyboardAction('.', actions.nextPost())
    testKeyboardAction('>', actions.nextPost())
    testKeyboardAction('N', actions.nextPost(), { shiftKey: true })
    testKeyboardAction('n', actions.nextPost(), { shiftKey: true })
  })

  it('should goto previousPost', () => {
    testKeyboardAction(',', actions.previousPost())
    testKeyboardAction('<', actions.previousPost())
    testKeyboardAction('P', actions.previousPost(), { shiftKey: true })
    testKeyboardAction('p', actions.previousPost(), { shiftKey: true })
  })

  it('should configTogglePlay', () => {
    testKeyboardAction(' ', actions.configTogglePlay())
    testKeyboardAction('Space', actions.configTogglePlay())
    testKeyboardAction('Enter', actions.configTogglePlay())
    testKeyboardAction('k', actions.configTogglePlay())
  })

  it('should configToggleAutoAdvance', () => {
    testKeyboardAction('a', actions.configToggleAutoAdvance())
  })

  it('should configEnableFullscreen', () => {
    testKeyboardAction('f', actions.configEnableFullscreen())
  })

  it('should configDisableFullscreen', () => {
    testKeyboardAction('Escape', actions.configDisableFullscreen())
  })

  it('should playerScanBackwards', () => {
    testKeyboardAction('ArrowLeft', actions.playerScanBackwards(5))
  })

  it('should playerScanForwards', () => {
    testKeyboardAction('ArrowRight', actions.playerScanForwards(10))
  })

  it('should playerVolumeUp', () => {
    testKeyboardAction('ArrowUp', actions.playerVolumeUp())
  })

  it('should playerVolumeDown', () => {
    testKeyboardAction('ArrowDown', actions.playerVolumeDown())
  })

  it('should playerJumpTo', () => {
    Array(9).fill().map((_, i) => (
      testKeyboardAction(i.toString(), actions.playerJumpTo(i))
    ))
  })

  it('should end', () => {
    testKeyboardAction('End', actions.playerJumpTo(10))
  })
})
