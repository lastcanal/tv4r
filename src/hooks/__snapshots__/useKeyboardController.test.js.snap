// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useKeyboardController handle heydown events 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ConnectedRouterWithContext)
    hashType="noslash"
    history={
      Object {
        "action": "POP",
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [MockFunction],
        "listen": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
            Array [
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "location": Object {},
        "push": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "replace": [MockFunction],
      }
    }
  >
    <ConnectedRouterWithContext
      hashType="noslash"
      history={
        Object {
          "action": "POP",
          "go": [MockFunction],
          "goBack": [MockFunction],
          "goForward": [MockFunction],
          "listen": [MockFunction] {
            "calls": Array [
              Array [
                [Function],
              ],
              Array [
                [Function],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
          "location": Object {},
          "push": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
          "replace": [MockFunction],
        }
      }
      onLocationChanged={[Function]}
    >
      <ConnectedRouter
        hashType="noslash"
        history={
          Object {
            "action": "POP",
            "go": [MockFunction],
            "goBack": [MockFunction],
            "goForward": [MockFunction],
            "listen": [MockFunction] {
              "calls": Array [
                Array [
                  [Function],
                ],
                Array [
                  [Function],
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "location": Object {},
            "push": [MockFunction] {
              "calls": Array [
                Array [
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                  },
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "replace": [MockFunction],
          }
        }
        onLocationChanged={[Function]}
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <Router
          history={
            Object {
              "action": "POP",
              "go": [MockFunction],
              "goBack": [MockFunction],
              "goForward": [MockFunction],
              "listen": [MockFunction] {
                "calls": Array [
                  Array [
                    [Function],
                  ],
                  Array [
                    [Function],
                  ],
                ],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": undefined,
                  },
                  Object {
                    "type": "return",
                    "value": undefined,
                  },
                ],
              },
              "location": Object {},
              "push": [MockFunction] {
                "calls": Array [
                  Array [
                    Object {
                      "hash": "",
                      "pathname": "/",
                      "search": "",
                    },
                  ],
                ],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": undefined,
                  },
                ],
              },
              "replace": [MockFunction],
            }
          }
        >
          <Wrapped
            dispatch={[Function]}
          >
            <div />
          </Wrapped>
        </Router>
      </ConnectedRouter>
    </ConnectedRouterWithContext>
  </Connect(ConnectedRouterWithContext)>
</Provider>
`;
